services/route_manager/src/main/java/com/futurewei/alcor/route/controller/RouterController.java

```diff
@@ -15,14 +15,287 @@
+            RestPreconditionsUtil.verifyResourceFound(projectid);
+
+            routetables = this.routerService.getVpcRouteTables(projectid, vpcid);
+
+        } catch (Exception e) {
+            throw e;
+        }
+
+        return new RouteTablesWebJson(routetables);
+    }
+
+    @RequestMapping(
+            method = GET,
+            value = {"/project/{projectid}/routetables/{routetableid}"})
+    @DurationStatistics
+    public RouteTableWebJson getVpcRouteTableById(@PathVariable String projectid, @PathVariable String routetableid) throws Exception {
+
+        RouteTable routetable = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(routetableid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyResourceFound(projectid);
+
+            routetable = this.routeTableDatabaseService.getByRouteTableId(routetableid);
+
+        } catch (ParameterNullOrEmptyException e) {
+            throw e;
+        }
+
+        return new RouteTableWebJson(routetable);
+    }
+
+    @RequestMapping(
+            method = GET,
+            value = {"/project/{projectid}/subnets/{subnetid}/routetable"})
+    @DurationStatistics
+    public RouteTableWebJson getOrCreateSubnetRouteTable(@PathVariable String projectid, @PathVariable String subnetid) throws Exception {
+
+        RouteTable routeTable = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(subnetid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyResourceFound(projectid);
+
+            routeTable = this.routerService.getOrCreateSubnetRouteTable(projectid, subnetid);
+
+        } catch (ParameterNullOrEmptyException e) {
+            throw e;
+        } catch (ExistMultipleSubnetRouteTable e) {
+            logger.log(Level.WARNING, e.getMessage() + " , subnetId: " + subnetid);
+            throw e;
+        } catch (DatabasePersistenceException e) {
+            throw e;
+        }
+
+        return new RouteTableWebJson(routeTable);
+    }
+
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/subnets/{subnetid}/routetable"})
+    @DurationStatistics
+    public RouteTableWebJson updateSubnetRouteTable(@PathVariable String projectid, @PathVariable String subnetid, @RequestBody RouteTableWebJson resource) throws Exception {
+
+        RouteTable routetable = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(subnetid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyResourceFound(projectid);
+
+            // check resource
+            if (!RouteManagerUtil.checkVpcDefaultRouteTableWebJsonResourceIsValid(resource)) {
+                throw new ResourceNotValidException("request resource is invalid");
+            }
```

Comment:
@kevin-zhonghao This function is for updateSubnetRouteTable, why the 'check resource' uses  `checkVpcDefaultRouteTableWebJsonResourceIsValid(resource)`

Reply:
That's a name mistake, it's already changed to checkSubnetRouteTableWebJsonResourceIsValid